testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs

```diff
@@ -211,6 +214,98 @@ async fn test_room_avatar_group_conversation() -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn test_joined_user_can_create_push_context_with_room_list_service() -> Result<()> {
+    let bob =
+        TestClientBuilder::new("bob".to_owned()).randomize_username().use_sqlite().build().await?;
+
+    let alice = TestClientBuilder::new("alice".to_owned())
+        .randomize_username()
+        .use_sqlite()
+        .build()
+        .await?;
+
+    // Set up regular sync for Alice to start with.
+    let a = alice.clone();
+    let alice_handle = spawn(async move {
+        if let Err(err) = a.sync(Default::default()).await {
+            warn!("alice sync errored: {err}");
+        }
+    });
+
+    // Set up standard sync for Bob.
+    let b = bob.clone();
+    let bob_handle = spawn(async move {
+        if let Err(err) = b.sync(Default::default()).await {
+            warn!("bob sync errored: {err}");
+        }
+    });
+
+    // Alice creates a room and invites Bob.
+    let alice_room = alice
+        .create_room(assign!(CreateRoomRequest::new(), {
+            invite: vec![bob.user_id().unwrap().to_owned()],
+            is_direct: false,
+        }))
+        .await?;
+
+    sleep(Duration::from_secs(1)).await;
+
+    bob.join_room_by_id(alice_room.room_id()).await?;
+
+    let alice_room = alice.get_room(alice_room.room_id()).unwrap();
+    assert_eq!(alice_room.state(), RoomState::Joined);
+
+    // Bob sends a message.
+    bob.get_room(alice_room.room_id())
+        .unwrap()
+        .send(RoomMessageEventContent::text_plain("hello world"))
+        .await?;
+
+    // Stop previous sync for Alice.
+    alice_handle.abort();
+
+    // Start a new sync using the room list service for another device of Alice.
+    let hs = alice.homeserver();
+    let sliding_sync_url = alice.sliding_sync_proxy();
+    let alice_id = alice.user_id().unwrap().localpart().to_owned();
+
+    let alice = Client::new(hs).await?;
+    alice.set_sliding_sync_proxy(sliding_sync_url);
+    alice.matrix_auth().login_username(&alice_id, &alice_id).await?;
+
+    let room_list_service = Arc::new(RoomListService::new(alice.clone()).await?);
+
+    let actions = Arc::new(Mutex::new(None));
+    let a = actions.clone();
+    alice.add_event_handler(|ev: OriginalSyncRoomMessageEvent, actions: Vec<Action>| async move {
+        let MessageType::Text(text_content) = ev.content.msgtype else { return };
+        if text_content.body == "hello world" {
+            *a.lock().await = Some(actions);
+        }
+    });
+
```

Comment:
I'd appreciate some comments near here explaining what we are actually testing here and what outcome we are expecting. I might even appreciate splitting some of this code out into function to make it easier to follow...

Reply:
Thanks for the review! I've added lots of comments, then realized I could simplify the test a bit. Can you take another look at the comments, and hit merge if you're happy with those, please?